fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/cursors/SizeStatisticsCollectorCursor.java

```diff
@@ -0,0 +1,527 @@
+        });
+    }
+
+    //update partial stats
+    private void updateStatistics(@Nonnull KeyValue kv) {
+        this.sizeStatisticsResults.keyCount += 1;
+        this.sizeStatisticsResults.keySize += kv.getKey().length;
+        this.sizeStatisticsResults.maxKeySize = Math.max(this.sizeStatisticsResults.maxKeySize, kv.getKey().length);
+        this.sizeStatisticsResults.valueSize += kv.getValue().length;
+        this.sizeStatisticsResults.maxValueSize = Math.max(this.sizeStatisticsResults.maxValueSize, kv.getValue().length);
+        int totalSize = kv.getKey().length + kv.getValue().length;
+        if (totalSize > 0) {
+            this.sizeStatisticsResults.sizeBuckets[Integer.SIZE - Integer.numberOfLeadingZeros(totalSize) - 1] += 1;
+        }
+    }
+
+    @Nonnull
+    @Override
+    public Executor getExecutor() {
+        return context.getExecutor();
+    }
+
+    @Nonnull
+    @Override
+    public boolean accept(@Nonnull RecordCursorVisitor visitor) {
+        visitor.visitEnter(this);
+        return visitor.visitLeave(this);
+    }
+
+    @Nonnull
+    private SizeStatisticsResults emitFinalStats() {
+        return sizeStatisticsResults;
+    }
+
+    @Nonnull
+    @Override
+    @Deprecated
+    public CompletableFuture<Boolean> onHasNext() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Nullable
+    @Override
+    @Deprecated
+    public SizeStatisticsResults next() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Nullable
+    @Override
+    @Deprecated
+    public byte[] getContinuation() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Nonnull
+    @Override
+    @Deprecated
+    public RecordCursor.NoNextReason getNoNextReason() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public void close() {
+        if (hasNextFuture != null) {
+            hasNextFuture.cancel(false);
+        }
+    }
+
+    //form a continuation that allows us to restart statistics aggregation from where we left off
+    private class SizeStatisticsCollectorCursorContinuation implements RecordCursorContinuation {
+        @Nonnull
+        private final RecordCursorResult<KeyValue> currentKvResult;
+        @Nonnull
+        private final Function<byte[], RecordCursorProto.SizeStatisticsContinuation> continuationFunction;
+        @Nonnull
+        private final boolean isEnd = false;
+        @Nullable
+        private byte[] cachedBytes;
+
+        private SizeStatisticsCollectorCursorContinuation(RecordCursorResult<KeyValue> currentKvResult) {
+            this.cachedBytes = null;
+            this.currentKvResult = currentKvResult;
+
+            //defer forming bytes until requested
+            continuationFunction = b -> {
+                RecordCursorProto.SizeStatisticsPartialResults.Builder pr;
+                if (finalResultsEmitted == false) {
+                    //more to aggregate put partial results in continuation
```

Comment:
```suggestion
                    //more to aggregate; put partial results in continuation
```

Reply:
yeah ok :)

Reply:
Never mind. Will clean up the logic to accommodate.